# Pokemon Database

## first we need to import the necessary libraries

In [1]:
import numpy as np
import pandas as pd

## put in a DataFrame all data from pokemon.csv

In [2]:
data = pd.read_csv('pokemon.csv')
data.head()

abilities  against_bug  against_dark  against_dragon  \
0  ['Overgrow', 'Chlorophyll']          1.0           1.0             1.0   
1  ['Overgrow', 'Chlorophyll']          1.0           1.0             1.0   
2  ['Overgrow', 'Chlorophyll']          1.0           1.0             1.0   
3     ['Blaze', 'Solar Power']          0.5           1.0             1.0   
4     ['Blaze', 'Solar Power']          0.5           1.0             1.0   

   against_electric  against_fairy  against_fight  against_fire  \
0               0.5            0.5            0.5           2.0   
1               0.5            0.5            0.5           2.0   
2               0.5            0.5            0.5           2.0   
3               1.0            0.5            1.0           0.5   
4               1.0            0.5            1.0           0.5   

   against_flying  against_ghost  ...  percentage_male  pokedex_number  \
0             2.0            1.0  ...             88.1               1   
1             2.0            1.0  ...             88.1               2   
2             2.0            1.0  ...             88.1               3   
3             1.0            1.0  ...             88.1               4   
4             1.0            1.0  ...             88.1               5   

   sp_attack  sp_defense  speed  type1   type2  weight_kg  generation  \
0         65          65     45  grass  poison        6.9           1   
1         80          80     60  grass  poison       13.0           1   
2        122         120     80  grass  poison      100.0           1   
3         60          50     65   fire     NaN        8.5           1   
4         80          65     80   fire     NaN       19.0           1   

   is_legendary  
0             0  
1             0  
2             0  
3             0  
4             0  

[5 rows x 41 columns]

## show all columns from the data

In [3]:
data.columns

Index(['abilities', 'against_bug', 'against_dark', 'against_dragon',
       'against_electric', 'against_fairy', 'against_fight', 'against_fire',
       'against_flying', 'against_ghost', 'against_grass', 'against_ground',
       'against_ice', 'against_normal', 'against_poison', 'against_psychic',
       'against_rock', 'against_steel', 'against_water', 'attack',
       'base_egg_steps', 'base_happiness', 'base_total', 'capture_rate',
       'classfication', 'defense', 'experience_growth', 'height_m', 'hp',
       'japanese_name', 'name', 'percentage_male', 'pokedex_number',
       'sp_attack', 'sp_defense', 'speed', 'type1', 'type2', 'weight_kg',
       'generation', 'is_legendary'],
      dtype='object')

## Create the tables to import to PostgreSQL

### The pokemons table will look like this:

| pokedex_number | name | hp | attack | defense | sp_attack| sp_defense | speed | fk_type | fk_classfication |
|----------------|------|----|--------|---------|----------|------------|-------|---------|------------------|
| 1 |	Bulbasaur | 45 | 49 | 49 | 65 | 65 | 45 | Seed Pokémon |
| 2 |	Ivysaur |	60 | 62 |	63 | 80 | 80 | 60 | Seed Pokémon |
| 3 |	Venusaur | 80 | 100 | 123 | 122 | 120 | 80 | Seed Pokémon |
| 4 |	Charmander | 39 | 52 | 43 | 60 | 50 | 65 | Lizard Pokémon |
| 5 |	Charmeleon | 58 | 64 | 58 | 80 | 65 |	80 | Flame Pokémon |

In [4]:
pokemon = data[['pokedex_number', 'name', 'hp', 
                'attack', 'defense', 'sp_attack',
                'sp_defense', 'speed', 'classfication',
                'generation']]
pokemon.head()

pokedex_number        name  hp  attack  defense  sp_attack  sp_defense  \
0               1   Bulbasaur  45      49       49         65          65   
1               2     Ivysaur  60      62       63         80          80   
2               3    Venusaur  80     100      123        122         120   
3               4  Charmander  39      52       43         60          50   
4               5  Charmeleon  58      64       58         80          65   

   speed   classfication  generation  
0     45    Seed Pokémon           1  
1     60    Seed Pokémon           1  
2     80    Seed Pokémon           1  
3     65  Lizard Pokémon           1  
4     80   Flame Pokémon           1

### The pokemon classfication table will look like this:

| classficationID | classfication |
|-----------------|---------------|
| 1 | Abundance Pokémon |
| 2 | Acorn Pokémon |
| 3 | Alpha Pokémon |
| 4 | Angler Pokémon |
| 5 | Ant Pit Pokémon |


In [5]:
classID = pd.Series(range(1, len(set(data.classfication))+1))

classfications = pd.DataFrame(set(data.classfication), columns=['classfication'])
classfications.sort_values(by=['classfication'], inplace=True)
classfications.set_index(classID, inplace=True)
classfications.reset_index(inplace=True)
classfications = classfications.rename(columns = {'index':'classficationID'})
classfications.to_csv('data/classfications.csv', index=False)
classfications.head()

classficationID      classfication
0                1  Abundance Pokémon
1                2      Acorn Pokémon
2                3      Alpha Pokémon
3                4     Angler Pokémon
4                5    Ant Pit Pokémon

### The pokemon type table will look like this:

| typeID | type |
|--------|------|
| 1 |	bug | 
| 2 |	dark | 
| 3 |	dragon | 
| 4 |	electric | 
| 5 |	fairy | 

In [6]:
typeID = pd.Series(range(1, len(set(data.type1))+1))

types = pd.DataFrame(set(data.type1), columns=['type'])
types.sort_values(by=['type'], inplace=True)
types.set_index(typeID, inplace=True)
types.reset_index(inplace=True)
types = types.rename(columns = {'index':'typeID'})
types.to_csv('data/types.csv', index=False)
types.head()

typeID      type
0       1       bug
1       2      dark
2       3    dragon
3       4  electric
4       5     fairy

### The pokemon abilities table will look like this:

| abilityID | ability |
|-----------|---------|

In [11]:
abilities = data[['abilities']]
abilities.to_csv('data/abilities.csv', index=False)
abilities.explode('abilities').sort_values(by=['abilities'])
abilities

abilities
0    ['Overgrow', 'Chlorophyll']
1    ['Overgrow', 'Chlorophyll']
2    ['Overgrow', 'Chlorophyll']
3       ['Blaze', 'Solar Power']
4       ['Blaze', 'Solar Power']
..                           ...
796              ['Beast Boost']
797              ['Beast Boost']
798              ['Beast Boost']
799              ['Prism Armor']
800               ['Soul-Heart']

[801 rows x 1 columns]

SyntaxError: invalid syntax (3330908697.py, line 1)